### PR TITLE
docs: clarify closure usage in gan example

### LINKED ISSUE #7337: Tutorial "step optimizer at arbitrary intervals" does not call train_step every iteration
## 🐛 Bug

The tutorial [step optimizer at arbitrary intervals](https://pytorch-lightning.readthedocs.io/en/latest/common/optimizers.html#step-optimizers-at-arbitrary-intervals) has a bug/feature(*) that the closure is evaluated (with `optimizer_idx=1`) only every other batch, not every batch. 

(*bug/feature) Note: this behavior is intended for e.g. GAN training where a part of the model is only trained every n batches, and the tutorial has the expected behavior. But the language *looks like* the tutorial is also applicable to people who are trying to implement `accumulate_grad_batches` here (Pytorch-lightning forbids us using `accumulate_grad_batches` if we override `optimizer_step`). For this use case, using the tutorial leads to the unexpected behavior.

### To Reproduce

[Colab](https://colab.research.google.com/drive/1xlINNCJ1ugRp726ux2h5Ya3NvTj2O09-?usp=sharing)

Difference: overrode `optimizer_step`, added printing in `training_step`.

### Expected behavior

The training step is called every batch.
```
training_step(batch_idx=0)
training_step(batch_idx=1)
training_step(batch_idx=2)
training_step(batch_idx=3)
training_step(batch_idx=4)
training_step(batch_idx=5)
training_step(batch_idx=6)
...
```
### Actual behavior
```
training_step(batch_idx=1)
training_step(batch_idx=3)
training_step(batch_idx=5)
training_step(batch_idx=7)
training_step(batch_idx=9)
training_step(batch_idx=11)
training_step(batch_idx=13)
...
```

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #7337

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: docs/source/common/optimizers.rst [hunk 1]
```diff
@@ -397,6 +397,8 @@ For example, here step optimizer A every batch and optimizer B every 2 batches.
         if optimizer_idx == 1:
             if (batch_idx + 1) % 2 == 0:
                 optimizer.step(closure=optimizer_closure)
+            else:
+                # closure/training_step won't be called in this case
```
Comment (Borda) [inline docs/source/common/optimizers.rst:401]:
```suggestion
                ...  # closure/training_step won't be called in this case
```

**Review by Borda — APPROVED** [review]

File: docs/source/common/optimizers.rst (hunk 1 above)
Comment (kaushikb11) [reply docs/source/common/optimizers.rst:401]: Instead of adding an `else`, why not add `closure/training_step would only be called in this case` comment under if statement?

Comment (awaelchli) [reply docs/source/common/optimizers.rst:401]: no specific reason, one or the other way is equivalent in the information it provides.
doing both is probably redundant

Comment (awaelchli) [inline docs/source/common/optimizers.rst:401]:
```suggestion
            if (batch_idx + 1) % 2 == 0:
                # closure/training_step will ONLY be called in this case
                optimizer.step(closure=optimizer_closure)
                
```

File: docs/source/common/optimizers.rst [hunk 2]
```diff
@@ -396,7 +396,9 @@ For example, here step optimizer A every batch and optimizer B every 2 batches.
         # update discriminator every 2 steps
         if optimizer_idx == 1:
             if (batch_idx + 1) % 2 == 0:
+                # closure/training_step will ONLY be called in this case
                 optimizer.step(closure=optimizer_closure)
+
```
Comment (awaelchli) [inline docs/source/common/optimizers.rst:401]:
```suggestion
                optimizer.step(closure=optimizer_closure)
```

File: docs/source/common/optimizers.rst [hunk 3]
```diff
@@ -396,6 +396,7 @@ For example, here step optimizer A every batch and optimizer B every 2 batches.
         # update discriminator every 2 steps
         if optimizer_idx == 1:
             if (batch_idx + 1) % 2 == 0:
+                # closure/training_step will ONLY be called in this case
```
Comment (carmocca) [inline docs/source/common/optimizers.rst:399]:
```suggestion
                # The closure (which includes the `training_step`) won't run if this isn't called
```

**Review by carmocca — APPROVED** [review]

File: docs/source/common/optimizers.rst (hunk 3 above)
Comment (awaelchli) [reply docs/source/common/optimizers.rst:399]: no, here it is getting called. 
the point of this is to say there is a missing "else" branch in this if clause where we intentionally omit the call to the step.

Comment (carmocca) [reply docs/source/common/optimizers.rst:399]:
```suggestion
                # The closure (which includes the `training_step`) won't run if the line below isn't executed
```

clearer?

Comment (awaelchli) [reply docs/source/common/optimizers.rst:399]: I would have preferred the else approach initially, but let's not get distracted by these details

File: docs/source/common/optimizers.rst [hunk 4]
```diff
@@ -396,6 +396,7 @@ For example, here step optimizer A every batch and optimizer B every 2 batches.
         # update discriminator every 2 steps
         if optimizer_idx == 1:
             if (batch_idx + 1) % 2 == 0:
+                # The closure (which includes the `training_step`) won't run if the line below isn't executed
```
Comment (awaelchli) [inline docs/source/common/optimizers.rst:399]:
```suggestion
                # the closure (which includes the `training_step`) won't run if the line below isn't executed
```

**Review by tchaton — APPROVED** [review]
LGTM !
